### ISSUE TITLE
typecheck returns `undeclared name` for packages not matching import string

### ISSUE THREAD

**Issue by mbrancato** [golangci/golangci-lint#3605]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

When importing a package like `gopkg.in/yaml.v3` which has a base package of `yaml` is detected as undeclared.

Other than an import alias, is there a way to avoid this? Go compiles this just fine.

```
myfile_test.go:12:40: undeclared name: `yaml` (typecheck)
func (e *YExample) UnmarshalYAML(node *yaml.Node) error {
                                       ^
myfile_test.go:21:9: undeclared name: `yaml` (typecheck)
        err := yaml.Unmarshal([]byte(data), &e)
```

### Version of golangci-lint

<details>

```console
% golangci-lint --version
golangci-lint has version 1.51.1 built from b87d2c1 on 2023-02-05T00:48:10Z

```

</details>

### Configuration file

none

### Go environment

<details>

```console
% go version && go env
go version go1.18.5 darwin/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/mike/Library/Caches/go-build"
GOENV="/Users/mike/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/mike/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/mike/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go@1.18/1.18.5/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go@1.18/1.18.5/libexec/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18.5"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/8l/0qhz2jl149bdfk71_vsx2t6r0000gn/T/go-build148219607=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

</details>

### Verbose output of running

<details>

```console
% golangci-lint run --disable-all --enable typecheck -v myfile_test.go
INFO [config_reader] Config search paths: [./ /Users/mike /Users /] 
INFO [lintersdb] Active 1 linters: [typecheck]    
INFO [loader] Go packages loading at mode 575 (compiled_files|exports_file|name|deps|files|imports|types_sizes) took 395.504895ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 2.311198ms 
INFO [linters_context/goanalysis] analyzers took 50.633µs with top 10 stages: typecheck: 50.633µs 
INFO [runner] Processors filtering stat (out/in): max_from_linter: 2/2, path_shortener: 2/2, filename_unadjuster: 2/2, path_prettifier: 2/2, skip_files: 2/2, max_per_file_from_linter: 2/2, exclude-rules: 2/2, nolint: 2/2, diff: 2/2, source_code: 2/2, skip_dirs: 2/2, autogenerated_exclude: 2/2, identifier_marker: 2/2, exclude: 2/2, severity-rules: 2/2, max_same_issues: 2/2, path_prefixer: 2/2, cgo: 2/2, uniq_by_line: 2/2, sort_results: 2/2 
INFO [runner] processing took 1.196406ms with stages: nolint: 737.649µs, source_code: 280.279µs, path_prettifier: 67.116µs, identifier_marker: 37.274µs, exclude-rules: 22.872µs, path_shortener: 14.586µs, cgo: 12.716µs, skip_dirs: 11.042µs, max_from_linter: 3.384µs, uniq_by_line: 3.226µs, max_same_issues: 2.17µs, filename_unadjuster: 840ns, autogenerated_exclude: 792ns, max_per_file_from_linter: 681ns, skip_files: 396ns, diff: 367ns, exclude: 339ns, severity-rules: 301ns, sort_results: 243ns, path_prefixer: 133ns 
myfile_test.go:12:40: undeclared name: `yaml` (typecheck)
func (e *YExample) UnmarshalYAML(node *yaml.Node) error {
                                       ^
myfile_test.go:21:9: undeclared name: `yaml` (typecheck)
        err := yaml.Unmarshal([]byte(data), &e)
               ^
INFO File cache stats: 1 entries of total size 356B 
INFO Memory: 6 samples, avg is 32.4MB, max is 40.4MB 
INFO Execution took 470.262561ms                       
```

</details>

### Code example or link to a public repository

<details>

```go
package mypkg

import (
	"gopkg.in/yaml.v3"
	"testing"
)

type YExample struct {
	Attr1 int `json:"attr1" yaml:"attr1"`
}

func (e *YExample) UnmarshalYAML(node *yaml.Node) error {
	return nil
}

func TestYExample(t *testing.T) {
	var e YExample
	data := `
attr1: 5
`
	err := yaml.Unmarshal([]byte(data), &e)
	if err != nil {
		t.Error(err)
	}
}

```

</details>

**Comment by mbrancato** [comment]
I noticed this also happens on imports for package versions >= `v2`. Go doesn't need `v0` or `v1` specifiers, but it does for `v2` and above. 

e.g. `github.com/scylladb/gocqlx/v2`

```
cql/cql.go:54:24: undeclared name: `gocqlx` (typecheck)
func createSession() (*gocqlx.Session, error) {
```

However importing like this:
```go
	gocqlx "github.com/scylladb/gocqlx/v2"
```

Will cause some in-editor checks to complain about redundant aliases. 

```
Redundant alias 
 Inspection info: Reports aliases of imported packages that may be omitted.
Usually, such aliases equal to the names of the imported packages, so aliases have no effect and one can use package names directly.
Example:
import fmt "fmt"
The fmt alias duplicates the package name that is also named "fmt". To delete the alias, use the Delete import alias quick-fix.
After the quick-fix is applied:
import "fmt"
```

**Comment by mymmrac** [comment]
Started getting same error (`undeclared name: <package-name> (typecheck)` after updating from Go 1.19 to 1.20 without update of linter, is it expected?

**Comment by ldez** [comment]
go1.20 is only supported by golangci-lint > v1.51.0

**Comment by mymmrac** [comment]
Oh, didn't notice that, thanks for info

**Comment by YTGhost** [comment]
I meet similar problem to @mbrancato 
e.g. `"k8s.io/klog/v2"`
```
pkg/yurthub/kubernetes/serializer/serializer.go:200:2: undeclared name: `klog` (typecheck)
        klog.Infof("%#+v is not found in client-go runtime scheme", gvr)
```
```
golangci-lint has version v1.51.2 built from (unknown, mod sum: "h1:yIcsT1X9ZYHdSpeWXRT1ORC/FPGSqDHbHsu9uk4FK7M=") on (unknown)
```

**Comment by ldez** [comment]
Same question so same answer.

**Comment by BlueBlue-Lee** [comment]
Same problem. How to resolve this? Is there any solution?
```
$ go version
go version go1.19.7 linux/amd64

golangci-lint has version v1.52.2 built with go1.19.7 from (unknown, mod sum: "h1:FrPElUUI5rrHXg1mQ7KxI1MXPAw5lBVskiz7U7a8a1A=") on (unknown)
```

**Comment by lzap** [comment]
For googlers, this can happen if you have a different version of go on your PATH. For example, in our project we build and lint with go1.19.6 which we all install with a script (`go install 

The problem is that [golang-lint detects](https://github.com/golangci/golangci-lint/blob/master/pkg/goutil/env.go) GOROOT and GOCACHE env variables by calling `go env` and it seems it ignores these variables when set:

```
# which go
/usr/bin/go

# GOROOT=/home/lzap/sdk/go1.19.6 GOCACHE=/home/lzap/.cache/go-build ./bin/golangci-lint run
... error appears...
```

In order to achieve what I want, I need to set PATH accordingly:

```
# PATH=/home/lzap/sdk/go1.19.6/bin ./bin/golangci-lint run
... no errors ...
```

This solved the problem for me. Pay attention when you are switching various Go versions either by environmental variables or PATH or other tools like direnv. What we do on our project is couple of make targets which install specific version of Go and golang-lint into local directory which is in `.gitignore`:

```
  PROJECT_DIR:=$(shell dirname $(abspath $(firstword $(MAKEFILE_LIST)))) 
  GO_VERSION?=1.19.6
  GOBIN:=$(PROJECT_DIR)/bin
  GO?=$(shell go env GOPATH)/bin/go$(GO_VERSION)
  GOROOT?=$(shell $(GO) env GOROOT)
  GOCACHE?=$(shell $(GO) env GOCACHE)
  
  GOLINT?=$(GOBIN)/golangci-lint
  .PHONY: install-go
  install-go: ## Install required Go version
          go install golang.org/dl/go$(GO_VERSION)@latest
          $(GO) download

  .PHONY: install-tools
  install-tools: ## Install required Go commands into ./bin
          @# NOTE: also update .github/workflows version
          GOBIN=$(GOBIN) $(GO) install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.52.2

  .PHONY: lint
  lint: ## Run Go language linter `golangci-lint`
          GOROOT=$(GOROOT) PATH=$(GOROOT)/bin:${PATH} GOCACHE=$(GOCACHE) $(GOLINT) run
```

I should probably remove GOROOT/GOCACHE variables being passed into the linter because it just ignores it according to my findings. Hopefully this was useful, cheers!
